File: services/business-services/business-menu-option-v3/infrastructure/mapper/menu_option_mapper.go
package mapper

import (
	"business-menu-option-v3/domain/model/request"
	"business-menu-option-v3/domain/model/response"
	"business-menu-option-v3/infrastructure/repository/entity"
	"github.com/mitchellh/mapstructure"
)

func RequestToEntity(request request.MenuOptionRequest) entity.MenuOption {
	var menuOption entity.MenuOption
	mapstructure.Decode(request, &menuOption)
	return menuOption
}

func UpdateRequestToEntity(request request.MenuOptionRequest, menuOption *entity.MenuOption) {
	mapstructure.Decode(request, menuOption)
}

func EntityToResponse(menuOption entity.MenuOption) response.MenuOptionResponse {
	var response response.MenuOptionResponse
	mapstructure.Decode(menuOption, &response)
	return response
}

File: application/backend/business/menu-v1/src/main/java/com/demo/bbq/business/menu/infrastructure/mapper/MenuOptionMapper.java
package com.demo.bbq.business.menu.infrastructure.mapper;

import com.demo.bbq.business.menu.domain.model.request.MenuOptionSaveRequest;
import com.demo.bbq.business.menu.domain.model.request.MenuOptionUpdateRequest;
import com.demo.bbq.business.menu.domain.model.response.MenuOption;
import com.demo.bbq.business.menu.infrastructure.repository.database.entity.MenuOptionEntity;
import com.demo.bbq.business.menu.infrastructure.repository.restclient.dto.ProductDto;
import com.demo.bbq.business.menu.infrastructure.repository.restclient.dto.ProductSaveRequestDto;
import com.demo.bbq.business.menu.infrastructure.repository.restclient.dto.ProductUpdateRequestDto;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;

@Mapper(componentModel = "spring")
public interface MenuOptionMapper {

  MenuOptionEntity fromSaveRequestToEntity(MenuOptionSaveRequest menuOption);

  MenuOptionEntity fromUpdateRequestToEntity(MenuOptionUpdateRequest menuOption, String productCode);

  @Mapping(target = "productCode", source = "product.code")
  MenuOption fromProductToResponse(MenuOptionEntity menuOption, ProductDto product);

  @Mapping(target = "code", source = "menuOption.productCode")
  ProductSaveRequestDto fromSaveRequestToProduct(MenuOptionSaveRequest menuOption, String scope);

  ProductUpdateRequestDto fromUpdateRequestToProduct(MenuOptionUpdateRequest menuOption, String scope);
}
